### PR TITLE
Add a gRPC client connector

### PR DESCRIPTION
**Pull request by aryan9600**
# Pull request

## Description

This PR adds a gRPC client connector. Tremor needs to be built with the protobuf information and with the grpc feature flag. For example: `PROTO_FILES="proto/helloworld/helloworld.proto,proto/routeguide/route_guide.proto" INCLUDE="proto/helloworld,proto/routeguide" cargo build --features grpc`

## Related

* Related Issues: affects #790 

## Checklist

* [x] The RFC, if required, has been submitted and approved
* [ ] Any user-facing impact of the changes is reflected in docs.tremor.rs
* [x] The code is tested
* [x] Use of unsafe code is reasoned about in a comment
* [ ] Update CHANGELOG.md appropriately, recording any changes, bug fixes, or other observable changes in behavior
* [ ] The performance impact of the change is measured (see below)

### CODE REVIEW

File: .cargo/config [hunk 1]
```diff
@@ -5,7 +5,7 @@ rustflags = ["-C", "target-feature=+avx,+avx2,+sse4.2"]
 rustflags = ["-C", "target-feature=+avx,+avx2,+sse4.2"]
 
 [target.aarch64-apple-darwin]
-rustflags = ["-C", "target-feature=+avx,+avx2,+sse4.2"]
```
Comment (mfelsche) [inline .cargo/config:8]: Good catch!

File: Cargo.toml [hunk 2]
```diff
@@ -206,8 +208,14 @@ testcontainers = "0.12"
 signal-hook = "0.3"
 signal-hook-async-std = "0.2"
 
+[build-dependencies]
+tremor-grpc-build = { git = "https://github.com/aryan9600/tremor-grpc-build", optional = true }
```
Comment (mfelsche) [inline Cargo.toml:212]: Can we pin this to a certain release tag?

File: src/connectors/impls/grpc/client.rs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    async fn connect(&mut self, _ctx: &ConnectorContext, _attempt: &Attempt) -> Result<bool> {
+        let grpc_client_handler = tremor_grpc::GrpcClientHandler::connect(
+            format!("{}:{}", self.config.host, self.config.port),
+            self.tx.clone(),
+        )
+        .await?;
+        if let Some(sink) = &mut self.sink {
+            let mut client_handler = sink.grpc_client_handler.lock().await;
```
Comment (mfelsche) [inline src/connectors/impls/grpc/client.rs:119]: sink and source now also have a `connect` method, so you can (and should) move this there, as the clone above will result in operating on two distinct entities

File: src/lib.rs [hunk 4]
```diff
@@ -14,7 +14,7 @@
 
 //! Tremor runtime
 
-#![deny(warnings)]
+// #![deny(warnings)]
```
Comment (mfelsche) [inline src/lib.rs:17]: can you comment this in? If this shall be changed, i'd recommend doing this in another PR. (It is totally fine for the Draft, though)

File: Cargo.toml [hunk 5]
```diff
@@ -175,6 +175,8 @@ tonic = { version = "0.5.2", default-features = false, features = [
   "transport",
   "tls",
 ] }
+prost = "0.8"
```
Comment (mfelsche) [inline Cargo.toml:178]: do we need those unconditionally, or can they be also only included when activating the `grpc` feature?

File: src/connectors.rs [hunk 6]
```diff
@@ -1159,6 +1159,8 @@ pub fn builtin_connector_types() -> Vec<Box<dyn ConnectorBuilder + 'static>> {
         Box::new(impls::crononome::Builder::default()),
         Box::new(impls::s3::Builder::default()),
         Box::new(impls::kafka::consumer::Builder::default()),
+        #[cfg(feature = "grpc")]
+        Box::new(impls::grpc::client::Builder::default())
```
Comment (mfelsche) [inline src/connectors.rs:1163]: Does this mean, we can only have 1 connector that runs any number of grpc services provided in the `PROTO_FILES` env var per compiled tremor binary? If we want to have, say, two different connectors from two different sets of proto files, would this be possible? 

Can you clarify what exactly can be run with tremor given the setup here? How is the grpc connector scoped?

What can we do to have different connector types per service? Or somehow group services on the command line? One connector type per proto file?

**Review by mfelsche — COMMENTED** [review]
First set of tired evening eyes glances over this draft. Looks good. One big question on the scope of the connector - will it handle all the services that were given in `PROTO_FILES` etc?

File: src/connectors.rs (hunk 6 above)
Comment (aryan9600) [reply src/connectors.rs:1163]: Right now, we have one connector that talks to all the services defined in all the provided files under `PROTO_FILES`. I don't know whether it'd be possible to have one connector per file, but it'd definitely require a lot of refactoring and restructuring.

**Pull request closed without merging**